### PR TITLE
WIP Enable leader election

### PR DESCRIPTION
**Pull request by sanchezl**
Enable leader election for migrator.

### CODE REVIEW

File: cmd/migrator/app/migrator.go [hunk 1]
```diff
@@ -77,6 +89,65 @@ func Run(ctx context.Context) error {
...
+	if hostname, err := os.Hostname(); err == nil {
+		id = hostname + "_" + id
+	}
+
+	if len(*resourceLockNamespace) == 0 {
+	}
+
+	lock := &resourcelock.ConfigMapLock{
+		ConfigMapMeta: metav1.ObjectMeta{
+			Name:      "migrator-lock",
```
Comment (deads2k) [inline cmd/migrator/app/migrator.go:131]: let's not be stingy: `kube-storage-version-migrator-lock`

File: cmd/migrator/app/migrator.go [hunk 2]
```diff
@@ -77,6 +89,65 @@ func Run(ctx context.Context) error {
...
+			Identity: id,
+		},
+	}
+
+	return &leaderelection.LeaderElectionConfig{
+		Lock:          lock,
+		LeaseDuration: 60 * time.Second,
+		RenewDeadline: 35 * time.Second,
+		RetryPeriod:   10 * time.Second,
+		Name:          "",
```
Comment (deads2k) [inline cmd/migrator/app/migrator.go:145]: what is this name plumbed through to?

File: cmd/migrator/app/migrator.go [hunk 3]
```diff
@@ -77,6 +89,65 @@ func Run(ctx context.Context) error {
...
+			Name:      "migrator-lock",
+			Namespace: *resourceLockNamespace,
+		},
+		Client: client,
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: id,
+		},
+	}
+
+	return &leaderelection.LeaderElectionConfig{
```
Comment (deads2k) [inline cmd/migrator/app/migrator.go:140]: be sure that we're gracefully releasing this lock on shutdown.

**Comment by sttts** [comment]
Any blocker here?

File: cmd/migrator/app/migrator.go [hunk 4]
```diff
@@ -77,6 +89,65 @@ func Run(ctx context.Context) error {
...
+		c.Run(ctx)
+	}
+
+	leaderelection.RunOrDie(ctx, *leaderElectionConfig)
+
 	panic("unreachable")
+
+}
+
+func getDefaultResourceLockNamespace() string {
```
Comment (mfojtik) [inline cmd/migrator/app/migrator.go:110]: is this something common we do in all operators?

File: cmd/migrator/app/migrator.go [hunk 5]
```diff
@@ -77,6 +89,65 @@ func Run(ctx context.Context) error {
...
+	if data, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace"); err == nil {
+		if ns := strings.TrimSpace(string(data)); len(ns) > 0 {
+			return ns
+		}
+	}
+	return "default"
```
Comment (mfojtik) [inline cmd/migrator/app/migrator.go:116]: i would maybe panic here if we missing the namespace file... but even better I would pass the namespace as flag?

File: cmd/migrator/app/migrator.go [hunk 6]
```diff
@@ -77,6 +89,65 @@ func Run(ctx context.Context) error {
...
+}
+
+func newLeaderElectionConfig(client v1.ConfigMapsGetter) *leaderelection.LeaderElectionConfig {
+
```
Comment (mfojtik) [inline cmd/migrator/app/migrator.go:120]: nit: remove this empty line

File: cmd/migrator/app/migrator.go (hunk 3 above)
Comment (mfojtik) [reply cmd/migrator/app/migrator.go:140]: i'm sure we don't

Comment (mfojtik) [reply cmd/migrator/app/migrator.go:140]: the context is not canceled ... i wonder why we have not used the command library we have in library-go

**Pull request closed without merging**